cmd/kubeadm/app/images/puller.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package images
+
+import (
+	"fmt"
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// ImagePuller is a struct that can pull images and hides the implementation (crictl vs docker)
+type ImagePuller struct {
+	criSocket  string
+	exec       utilsexec.Interface
+	crictlPath string
+}
+
+// NewImagePuller returns a ready to go ImagePuller
+func NewImagePuller(execer utilsexec.Interface, cfg *kubeadmapi.MasterConfiguration) (*ImagePuller, error) {
+	crictlPath, err := execer.LookPath("crictl")
+	if crictlPath == "" && err != nil {
+		return nil, err
+	}
+	return &ImagePuller{
+		exec:       execer,
+		criSocket:  cfg.GetCRISocket(),
+		crictlPath: crictlPath,
+	}, nil
+}
+
+// Pull pulls the actual image using either crictl or docker
+func (ip *ImagePuller) Pull(image string) error {
+	if ip.crictlPath == "" {
```

Comment:
This logic is wrong. We want to use docker if the cri socket is the default docker socket. Otherwise, use crictl